<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<!-- #BeginTemplate "/Templates/enhetssida.dwt" --><!-- #BeginEditable "doctitle" -->
  <title>ORFID</title>
<!-- #EndEditable -->
  <meta http-equiv="Content-Type"
 content="text/html; charset=iso-8859-1">
  <link rel="stylesheet" href="RTDocmall_files/profil.txt"
 type="text/css">
  <meta name="author" content="G&ouml;ran Sperber">
</head>
<body style="background-color: rgb(255, 255, 255); color: rgb(0, 0, 0);"
 class="rededge" leftmargin="0" topmargin="0" marginheight="0"
 marginwidth="0">
<table border="0" cellpadding="0" cellspacing="0" width="640">
  <tbody>
    <tr>
      <td rowspan="2" valign="top" width="160">
      <table bgcolor="#990000" border="0" cellpadding="0"
 cellspacing="0" height="401" width="100%">
        <tbody>
          <tr>
            <td height="135" valign="top" width="160"><a
 href="http://www.uu.se/"><img src="RTDocmall_files/uulogo_red160.gif"
 alt="Uppsala universitet" border="0" height="160" width="160"> </a> </td>
          </tr>
          <tr>
            <td class="sidomenyVIT" height="160" valign="top"><!-- #BeginEditable "index" -->
            <p>&nbsp;</p>
            <p><big><big><b>RetroTector</b></big></big></p>
            <p>&nbsp; </p>
<!-- #EndEditable --></td>
          </tr>
        </tbody>
      </table>
      </td>
      <td height="329" valign="top" width="480">
      <table class="normal" border="0" cellpadding="8" cellspacing="0"
 width="100%">
        <tbody>
          <tr>
            <td height="198" valign="top" width="464">
            <div class="normal">
            <div align="center">
            <h2>ORFID</h2>
            </div>
            <h2 align="center"><!-- #EndEditable --></h2>
<!-- #BeginEditable "text" -->
            <hr size="2" width="100%"><br>
ORFID is an <a href="Executors.html">Executor</a> which constructs
suggested proteins, "puteins", coded by ERVs found by <a
 href="RetroVID.html">RetroVID</a>. It depends on information about <a
 href="Motifs.html">Motif</a> hits found by RetroVID and transferred
through an ORFID <a href="ParameterFile.html">script</a> generated by
RetroVID, and it is therefore not normally meaningful to run it
interactively.<br>
            <br>
An <span style="font-weight: bold;">ordinary ORFID script</span>
specifies which <a href="Genes.html">Gene</a> and which <a
 href="VirusGenus.html">virus genus</a> are applicable, and ORFID
chooses an <a href="Alignments.html">Alignment</a> file accordingly.
The putein is constructed in quite close agreement with
the DNA; all ORFID does is occasionally to skip one or two bases to
cause a frame shift, and suggest where the putein starts and ends.<br>
            <br>
For ORFID to work, at least one of the Motif hits provided by RetroVID
must belong to a Motif that can be located in the Alignment. The putein
will be constructed in at least two segments; one leading
(up to the first located Motif hit); one trailing (after the last
located
Motif hit); and possibly more if there are more than one located Motif
hit.<br>
            <br>
Within each segment, the putein is constructed so as to maximize a
score, which for each acid is either (depending on which is larger):<br>
            <ul>
              <li>An 'aligned' score (max 1.0) dependent on the
similarity (according to the <a href="AcidMatrix.html">AcidMatrix</a>)
between the acid and the acid suggested by the Alignment at
this position.</li>
              <li><a name="UnAlignedScore"></a>An 'unaligned' score
(max 0.85) indicating how likely the reading frame of the acid is to be
a coding frame; it is calculated from several markers with variable
weigths (see below).</li>
            </ul>
In the case of the leading and trailing segments, a number of other
criteria related to the end points are also considered, such as<br>
            <ul>
              <li>score of <a href="PuteinStartMotif.html">PuteinStartMotif</a></li>
              <li>score of <a href="PuteinEndMotif.html">PuteinEndMotif<br>
                </a></li>
              <li>similarity to Kozak consensus</li>
              <li>proximity to hits by <a
 href="SpliceAcceptorMotif.html">SpliceAcceptorMotif</a></li>
              <li>proximity to hits by <a href="SiSeqMotif.html">SiSeqMotif</a></li>
              <li>proximity to hits by <a
 href="ProteaseCleavageMotif.html">ProteaseCleavageMotif</a></li>
              <li>proximity to hits by <a href="SlipperyMotif.html">SlipperyMotif</a></li>
              <li>proximity to hits by <a href="PseudoKnotMotif.html">PseudoKnotMotif</a></li>
              <li>stop codons</li>
            </ul>
The start and end points of the putein yielding the best scores
according to these criteria are chosen.<br>
            <br>
RetroVID provides ORFID with a list of all Motif hits in the relevant
Gene, in the parameter HitInfo. They may, however, not all be useful to
ORFID. In particular,
hits
with a low score may be false, which is not a big deal for RetroVID,
but
may reduce ORFID to gibberish. Therefore three parameters (ScoreFactor,
MinSSScore and MinHitScore) are related to weeding out of such
hits.
In principle, if the sum of the squared hit score is less than
MinSSScore,
ORFID
exits. If any Motif hit has a score lower than MinHitScore or lower
than
the average score times ScoreFactor, it is not used. If after this, no
hit
by a Motif in a <a href="MotifGroups.html">MotifGroup</a> with
an anchor in the Alignment is left, ORFID exits.<br>
            <br>
Normally, RetroVID generates the ordinary ORFID scripts unsweepable
(suffix 'Script.txt') and an <span style="font-weight: bold;">ORFID
"superscript"</span>, which lists
the ordinary scripts. It has a name without Gene specification and
suffix 'MScript_.txt' or 'NScript.txt'. When it is executed,
ORFID generates the puteins in parallell, so that their ends are
adjusted to each other according to a complex set of criteria.<br>
            <br>
Under unfavorable circumstances, in particular when Pol contains Motif
hits only in RT, ORFID may run out of memory, resulting in a
OutOfMemoryError message and/or an empty Putein file<br>
            <br>
The output of ORFID consists of:<br>
            <ul>
              <li>A Putein file, ie a <a href="Puteinview.html">Puteinview</a>
script
describing the putein, the Motif hits used in constructing it, the
reading frame and some indications about probable start and end of the
putein.</li>
              <li>Occasionally, a second Putein file with the same name
as the above, with an 'S' (for 'straightened') interspersed. It will be
similar to the above, but with a seemingly unnecessary deviation from
an ORF (which may or may not be due to imperfections in the Alignment
file) removed.<br>
              </li>
              <li>(Only if the Debugging option is on) An <a
 href="ORFIDview.html">ORFIDview</a> script (with name beginning
ORFIDout) with more detailed information, mainly useful when the putein
looks all screwy. The Putein file will contain information about all
the possible leading and trailing putein segments not used. This may
possibily lead to an OutOfMemoryError.<br>
              </li>
            </ul>
            <hr size="2" width="100%"><br>
The parameters of ORFID are:<br>
            <ul>
            </ul>
            <ul>
              <li>InputFile: Name of the file to get parameters from.
Only used in interactive mode.</li>
              <li>GeneScripts: Only if the script is a "superscript"
(see
above). A list of ordinary scripts.<br>
              </li>
              <li>Database: Database subdirectory to use.<br>
Default: Ordinary.<br>
              </li>
              <li>Eight parameters indicating the weighting when
calculating 'unaligned' score:</li>
              <ul>
                <li>NonAlignedScore: The initial value assigned to the
score for unaligned fitting. Default: 0.4.</li>
                <li>StopCodonFactor: Factor to multiply stop codon load
by in unaligned fitting. Default: -0.4.</li>
                <li>FrameShiftFactor: Penalty for frame shift in
unaligned fitting. Default: -1.5.</li>
                <li>ORFHexamerFactor: Factor to multiply <a
 href="Modifiers.html#ORFHexamerModifier">ORFHexamerModifier</a> by in
unaligned fitting. Default: 0.2.</li>
                <li>NonORFHexamerFactor: Factor to multiply <a
 href="Modifiers.html#NonORFHexamerModifier">NonORFHexamerModifier</a>
by in unaligned fitting. Default: -0.1.</li>
                <li>GlycosylationFactor: Factor to multiply <a
 href="Modifiers.html#GlycSiteModifier">GlycSiteModifier</a> by in
unaligned fitting of env. Default: 0.2.</li>
                <li>InORFBonus: Bonus to add if an acid is in a long
(&gt;50 acids) ORF. Default: 0.1.<br>
                </li>
                <li>MotifHitFactor: Factor to multiply score of Motif
hits by in unaligned fitting. Default: 0.2<br>
                  <br>
                </li>
              </ul>
              <li>MasterSkipPenalty. Penalty for skipping a filled
column in alignment. Default: -0.2<br>
              </li>
              <li>StopCodonValue. Value of stop codon in aligned
fitting. Default: -15.</li>
              <li>ScoreFactor: Factor for discarding lowscoring
MotifHits. Default: 0.1.</li>
              <li>MinSSScore: Lowest acceptable sum of squared MotifHit
scores.<br>
Default: 600.</li>
              <li>MinHitScore: Lowest acceptable individual MotifHit
score.<br>
Default: 15.<br>
              </li>
              <li>Gene: gag, pro, pol or env.</li>
              <li>VirusGenus: Any <a href="VirusGenus.html">virus
genus</a> symbol.</li>
              <li>DNAFile: The name of the file to read DNA from.
Default: Directory name + .txt</li>
              <li>Strand: Primary or Secondary. Default: Primary.</li>
              <li>FirstDNAStart: Calculated earliest possible start of
putein in DNA. Default: 1.</li>
              <li>LastDNAStart: Calculated latest possible start of
putein in DNA. Default: 1.</li>
              <li>FirstDNAEnd:&nbsp;Calculated earliest possible end of
putein in DNA. Default: 200000.</li>
              <li>LastDNAEnd: Calculated latest possible end of putein
in DNA. Default: 200000.</li>
              <li>HitInfo: A list of Motif hits. Each line contains:</li>
              <ul>
                <li>Hotspot position of the hit.</li>
                <li>The score of the hit.</li>
                <li>The Motif group of he Motif.</li>
                <li>The identifying string of the Motif class.</li>
                <li>The virus genus of the Motif.</li>
                <li>A string, typically the MOTIFSTRING of the Motif,
with the hotspot marked with '#'.</li>
                <li>The MOTIFORIGIN of the Motif.<br>
                </li>
              </ul>
              <li>OutputFile: Name of file for detailed output <a
 href="ORFIDview.html">(ORFIDview</a> script). Constructed from name of
script
or input file.</li>
              <li>PuteinFile: Name of file for putein output <a
 href="Puteinview.html">(Puteinview</a> script). Constructed from name
of
script or input file.</li>
              <li>ChainNumber: Number of parent Chain.</li>
              <li>Debugging: If this is Yes, a lot of extra information
is written to Standard Output, and an&nbsp;ORFIDview script is
generated. Default: No.</li>
              <li>Trouble: If present in the script, it describes some
problem (eg BrokenChain) encountered by RetroVID, suggesting that the
output
of ORFID may not be meaningful. It is copied to the putein file.<br>
              </li>
            </ul>
            <hr size="2" width="100%">
            <div align="right"><font size="-1"><small><a
 href="RetroTector.html"> To main RetroTector page</a> </small> </font><br>
            </div>
<!-- #EndEditable --></div>
            </td>
          </tr>
        </tbody>
      </table>
      </td>
    </tr>
    <tr>
      <td height="41" valign="top">
      <table border="0" cellpadding="8" cellspacing="0" width="100%">
        <tbody>
          <tr>
            <td height="41" valign="top" width="464">
            <hr color="#000000" noshade="noshade" size="1"><font
 face="verdana" size="-2"> &copy; 2002-2006 UPPSALA UNIVERSITET, Box
256, 751 05 Uppsala<br>
Responsible for this page: G&ouml;ran Sperber. Dept. of Neuroscience,
Unit of Physiology<br>
Goran.Sperber@neuro.uu.se</font><br>
            <font face="verdana" size="-2"> Last updated: 2006-11-02</font><!-- #EndDate --><!-- #EndTemplate --><br>
<!-- #EndDate --></td>
          </tr>
        </tbody>
      </table>
      </td>
    </tr>
  </tbody>
</table>
<!-- #EndTemplate --> <br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
</body>
</html>
