<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
<!-- #BeginTemplate "/Templates/enhetssida.dwt" --><!-- #BeginEditable "doctitle" -->
  <title>XonID</title>
<!-- #EndEditable -->
  <meta http-equiv="Content-Type"
 content="text/html; charset=iso-8859-1">
  <link rel="stylesheet" href="RTDocmall_files/profil.txt"
 type="text/css">
  <meta name="author" content="G&ouml;ran Sperber">
</head>
<body bgcolor="#ffffff" text="#000000" class="rededge" leftmargin="0"
 topmargin="0" marginwidth="0" marginheight="0">
<table width="640" border="0" cellpadding="0" cellspacing="0">
  <tbody>
    <tr>
      <td width="160" rowspan="2" valign="top">
      <table width="100%" border="0" cellpadding="0" cellspacing="0"
 height="401" bgcolor="#990000">
        <tbody>
          <tr>
            <td width="160" height="135" valign="top"><a
 href="http://www.uu.se/"><img src="RTDocmall_files/uulogo_red160.gif"
 width="160" height="160" border="0" alt="Uppsala universitet"> </a> </td>
          </tr>
          <tr>
            <td valign="top" height="160" class="sidomenyVIT"><!-- #BeginEditable "index" -->
            <p>&nbsp;</p>
            <p><big><big><b>RetroTector</b></big></big></p>
            <p>&nbsp; </p>
<!-- #EndEditable --></td>
          </tr>
        </tbody>
      </table>
      </td>
      <td width="480" valign="top" height="329">
      <table style="width: 100%;" border="0" cellpadding="8"
 cellspacing="0" class="normal">
        <tbody>
          <tr>
            <td
 style="width: 464px; vertical-align: top; height: 198px;">
            <div class="normal">
            <div align="center">
            <h2>XonID</h2>
            </div>
            <h2 align="center"><!-- #EndEditable --></h2>
<!-- #BeginEditable "text" -->
            <hr width="100%" size="2"><br>
XonID is an <a href="Executors.html">Executor</a> which searches for
traces of open reading frames not identified by <a href="ORFID.html">ORFID</a>,
avoiding the reading frame of a parallell Putein.
It uses methods similar to ORFID, but without use of <a
 href="AlignmentFiles.html">Alignments</a>. There are two strata of
findings:<br>
            <br>
            <ul>
              <li>XXons: these begin at a <a
 href="SpliceAcceptorMotif.html">SpliceAcceptorMotif</a> hit or at a
methionine codon at most 200 bases after one, and end at a stop codon
or a <a href="SpliceDonorMotif.html">SpliceDonorMotif</a> hit. XXons
have a reasonable chance of actually corresponding to exons.</li>
              <li>Xons: these have no particular requirements except
that they should have a low content of stop codons and frame shifts,
and avoid the reading frame of a parallell putein or XXon.
Xons are very vague entities which should <span
 style="font-weight: bold;">not</span> be taken seriously in the
absence of corroborating evidence, but may sometimes give useful hints.</li>
            </ul>
XonID is normally run using a script generated by <a
 href="RetroVID.html">RetroVID</a>. <span style="font-weight: bold;">It
will not run </span>if there are any <a
 href="ParameterFile.html#Sweepable">unswept</a> ORFID scripts in the
directory.<br>
            <br>
The output of XonID consists of a number of XXon- and Xon- files. These
are essentially <a href="Puteinview.html">Puteinview</a>
scripts, but without a number of not meaningful parameters.<br>
            <br>
            <hr width="100%" size="2"><br>
The parameters of XonID are:<br>
            <ul>
              <li>InputFile: Name of the file to get parameters from.
Only used in interactive mode.</li>
              <li>MinChainScore: If the parent Chain has lower score
than this, nothing is done. Default: 1000.</li>
              <li>LTRGagMinGap: If the gap between 5'LTR and Gag is at
least this much, XonID acts. Default: 100.</li>
              <li>GagProMinGap: If the gap between Gag and Pro is at
least this much, XonID acts. Default: 100.</li>
              <li>ProPolMinGap: If the gap between Pro and Pol is at
least this much, XonID acts. Default: 100.</li>
              <li>PolEnvMinGap: If the gap between Pol and Env is at
least this much, XonID acts. Default: 100.</li>
              <li>EnvLTRMinGap: If the gap between Env and 3'LTR is at
least this much, XonID acts. Default: 100.<br>
              </li>
              <li>Database: Database subdirectory to use.<br>
Default: Ordinary.</li>
              <li>ChainNumber: Number of parent Chain.</li>
              <li>FaultsPer51: A parameter regulating the fault
tolerance of the algorithm, mainly for XXons. The higher FaultsPer51,
the more tolerant. Default: 12.<br>
              </li>
              <li>SnakeThreshold: A parameter regulating the fault
tolerance of the algorithm for Xons. The higher SnakeThreshold, the
more tolerant. Default: 0.2.</li>
              <li>ChainsFile: Name of SelectedChains file to use.<br>
Default: 001SelectedChains.txt<br>
              </li>
              <li>Five parameters with similar roles as in ORFID:</li>
              <ul>
                <li>NonAlignedScore: The initial value assigned to the
score for unaligned fitting. Default: 0.4.</li>
                <li>StopCodonFactor: Factor to multiply stop codon load
by in unaligned fitting. Default: -0.4.</li>
                <li>FrameShiftFactor: Penalty for frame shift in
unaligned fitting. Default: -1.5.</li>
                <li>ORFHexamerFactor: Factor to multiply <a
 href="Modifiers.html#ORFHexamerModifier">ORFHexamerModifier</a> by in
unaligned fitting. Default: 0.2.</li>
                <li>NonORFHexamerFactor: Factor to multiply <a
 href="Modifiers.html#NonORFHexamerModifier">NonORFHexamerModifier</a>
by in unaligned fitting. Default: -0.1.<br>
                  <br>
                </li>
              </ul>
              <li>DNAFile: The name of the file to read DNA from.
Default: Directory name + .txt</li>
              <li>Strand: Primary or Secondary. Default: Primary.</li>
              <li>ChainStart: First position of search area in DNA.
Default: 1.</li>
              <li>ChainEnd: Last position of search area in DNA. 200000.</li>
            </ul>
            <hr width="100%" size="2">
            <div align="right"><font size="-1"><small><a
 href="RetroTector.html"> To main RetroTector page</a> </small> </font><br>
            </div>
<!-- #EndEditable --></div>
            </td>
          </tr>
        </tbody>
      </table>
      </td>
    </tr>
    <tr>
      <td height="41" valign="top">
      <table width="100%" border="0" cellpadding="8" cellspacing="0">
        <tbody>
          <tr>
            <td width="464" height="41" valign="top">
            <hr size="1" color="#000000" noshade="noshade"><font
 face="verdana" size="-2"> &copy; 2002-2005 UPPSALA UNIVERSITET, Box
256, 751 05 Uppsala<br>
Responsible for this page: G&ouml;ran Sperber. Dept. of Neuroscience,
Unit of Physiology<br>
Goran.Sperber@neuro.uu.se</font><br>
            <font face="verdana" size="-2"> Last updated: 2005-05-15</font><!-- #EndDate --><!-- #EndTemplate --><br>
<!-- #EndDate --></td>
          </tr>
        </tbody>
      </table>
      </td>
    </tr>
  </tbody>
</table>
<!-- #EndTemplate --> <br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
<br>
</body>
</html>
